### PR TITLE
Implement welcome email sending

### PR DESCRIPTION
**Pull request by lgonczar**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RPE-80

### Change description ###

Create a service for sending welcome emails. This includes a helper service for converting
private beta registrations into emails.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/pbis/EmailCreator.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.stream.Collectors;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.pbis.config.ApplicationConfig;
+import uk.gov.hmcts.reform.pbis.model.EmailTemplateMapping;
+import uk.gov.hmcts.reform.pbis.model.EmailToSend;
+import uk.gov.hmcts.reform.pbis.model.PrivateBetaRegistration;
+
+@Component
+public class EmailCreator {
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/pbis/EmailCreator.java]: Any ideas for a better name?

File: src/main/java/uk/gov/hmcts/reform/pbis/model/EmailToSend.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package uk.gov.hmcts.reform.pbis.model;
+
+import java.util.Map;
+
+public final class EmailToSend {
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/pbis/model/EmailToSend.java]: I've prepared our internal models as immutable classes with no accessors. As this may be controversial for some people, I can add getters. Please let me know if that's the case.

File: src/main/java/uk/gov/hmcts/reform/pbis/EmailCreator.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            registration.emailAddress,
+            getTemplateId(registration.service),
+            getFieldsForTemplate(registration),
+            registration.referenceId
+        );
+    }
+
+    private String getTemplateId(final String service) {
+        return emailTemplateMappings.get(service).getTemplateId();
+    }
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/pbis/EmailCreator.java]: looks like this function is only used once

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/pbis/EmailCreator.java]: also, is this null safe?

File: src/main/java/uk/gov/hmcts/reform/pbis/EmailService.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.pbis.model.EmailToSend;
+import uk.gov.hmcts.reform.pbis.model.PrivateBetaRegistration;
+import uk.gov.hmcts.reform.pbis.notify.NotificationClientProvider;
+import uk.gov.service.notify.NotificationClientApi;
+import uk.gov.service.notify.NotificationClientException;
+
+@Component
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/pbis/EmailService.java]: perhaps `@Service`?

File: src/main/java/uk/gov/hmcts/reform/pbis/EmailService.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            "Sending welcome email. Reference ID: %s", privateBetaRegistration.referenceId
+        ));
+
+        try {
+            EmailToSend emailToSend = emailCreator.createEmailToSend(privateBetaRegistration);
+
+            NotificationClientApi notificationClient =
+                notificationClientProvider.getClient(privateBetaRegistration.service);
+
+            sendEmail(emailToSend, notificationClient);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/pbis/EmailService.java]: inline `emailToSend` and `notificationClient`

File: src/main/java/uk/gov/hmcts/reform/pbis/model/EmailToSend.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    public final String emailAddress;
+    public final String templateId;
+    public final Map<String, String> templateFields;
+    public final String referenceId;
+
+
+    public EmailToSend(final String emailAddress,
+                       final String templateId,
+                       final Map<String, String> templateFields,
+                       final String referenceId) {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/pbis/model/EmailToSend.java]: let's keep formatting consistent

File: src/main/java/uk/gov/hmcts/reform/pbis/EmailCreator.java (hunk 3 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/pbis/EmailCreator.java]: It's used only once, but the call reads a bit better when it simply says 'give me the template for this service'.

It's safe - the public method starts with a check and if the service is absent, it throws an exception indicating a configuration problem

File: src/main/java/uk/gov/hmcts/reform/pbis/model/EmailToSend.java (hunk 6 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/pbis/model/EmailToSend.java]: do you mean the first line?

Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/pbis/model/EmailToSend.java]: done

File: src/main/java/uk/gov/hmcts/reform/pbis/EmailService.java (hunk 4 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/pbis/EmailService.java]: done

File: src/main/java/uk/gov/hmcts/reform/pbis/EmailCreator.java (hunk 3 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/pbis/EmailCreator.java]: ? `emailTemplateMappings` is a map

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/pbis/EmailCreator.java]: Another reason to inline it then. This private method is null safe. Just happens to never throw NPE because the code using it does the check.

Just noticed that this check (`assertTemplateConfiguredForService`) is also called only once.

Please inline both.

File: src/test/java/uk/gov/hmcts/reform/pbis/EmailServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        throws NotificationClientException {
+
+        doThrow(mock(NotificationClientException.class))
+            .when(notificationClient)
+            .sendEmail(anyString(), anyString(), anyMap(), anyString());
+
+        emailService.sendWelcomeEmail(privateBetaRegistration);
+    }
+
+    @Test(expected = EmailSendingException.class)
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/pbis/EmailServiceTest.java]: Did you consider assertJ way of testing exceptions?

Not insisting on it, but check it out if you haven't.

File: src/test/java/uk/gov/hmcts/reform/pbis/EmailServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    public void sendWelcomeEmail_should_throw_exception_when_client_provider_fails()
+        throws NotificationClientException {
+
+        doThrow(ServiceNotFoundException.class)
+            .when(notificationClientProvider)
+            .getClient(anyString());
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/pbis/EmailServiceTest.java]: how about
```
given(notificationClientProvider.getClient(anyString()).willThrow(...)
```
?

File: src/main/java/uk/gov/hmcts/reform/pbis/EmailCreator.java (hunk 3 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/pbis/EmailCreator.java]: There's nothing wrong with a method being called once. It's much more readable when a method's code deals with a certain level of details and you don't have to read through the parts you're not interested in, in order to find out what's happening.

If `createEmailToSend` contained the details of `assertTemplateConfiguredForService` it would just be less readable. Same with `getTemplateId`.

Just like we validate models once and don't check their correctness every time we use them, not all methods have to validate their inputs. Besides, this edge case is covered by unit tests.

File: src/test/java/uk/gov/hmcts/reform/pbis/EmailServiceTest.java (hunk 7 above)
Comment (lgonczar) [reply src/test/java/uk/gov/hmcts/reform/pbis/EmailServiceTest.java]: I didn't know about `catchThrowable` when writing this. Will change.

File: src/test/java/uk/gov/hmcts/reform/pbis/EmailServiceTest.java (hunk 8 above)
Comment (lgonczar) [reply src/test/java/uk/gov/hmcts/reform/pbis/EmailServiceTest.java]: I can change it, but is it worth it?

Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/pbis/EmailServiceTest.java]: imho it reads better and fits into given-when-then. Up to you.
